taskcluster/app_services_taskgraph/target_tasks.py

```diff
@@ -43,3 +45,41 @@ def filter(task):
                 and task.attributes.get('release-type') != 'release-only')
 
     return [l for l, task in full_task_graph.tasks.items() if filter(task)]
+
+
+def filter_release_promotion(
+    full_task_graph, filtered_for_candidates, shipping_phase
+):
+    def filter(task):
+        # Include promotion tasks; these will be optimized out
+        if task.label in filtered_for_candidates:
+            return True
+
+        if task.attributes.get("shipping_phase") == shipping_phase:
+            return True
+
+    return [
+        label for label, task in full_task_graph.tasks.items()
+        if filter(task)
+    ]
+
+
+@_target_task("promote")
+def target_tasks_promote(full_task_graph, parameters, graph_config):
+    return filter_release_promotion(
+        full_task_graph,
+        filtered_for_candidates=[],
+        shipping_phase="promote",
+    )
+
+
+@_target_task("ship")
+def target_tasks_ship(full_task_graph, parameters, graph_config):
+    filtered_for_candidates = target_tasks_promote(
+        full_task_graph,
+        parameters,
+        graph_config,
+    )
+    return filter_release_promotion(
+        full_task_graph, filtered_for_candidates, shipping_phase="ship"
+    )
```

Comment:
Does this mean that we'll run the `promote` tasks again in the `ship` phase or will they be cached somehow?